cdap-security/src/main/java/io/cdap/cdap/security/auth/AESCipher.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.security.auth;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.PBEKeySpec;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * AES cipher that supports encryption and decryption of data.
+ */
+public class AESCipher {
+  private static final String AES = "AES";
+  private static final String ENCRYPTION_ALGORITHM = "AES/CBC/PKCS5PADDING";
```

Comment:
nit: CBC mode for AES does not provide integrity which means an attacker can modify the ciphertext and potentially create another valid ciphertext. Can we instead switch to using an authenticated encryption mode of operation such as [Galois Counter Mode (GCM)](https://docs.oracle.com/javase/7/docs/technotes/guides/security/StandardNames.html#Cipher)? (This is a bit lower risk because changing a bit in the user credential should not violate security requirements)

Reply:
Done